File: app/src/main/java/com/elementary/tasks/core/views/BeforePickerView.java
package com.elementary.tasks.core.views;

import android.content.Context;
import android.content.res.TypedArray;
import androidx.annotation.Nullable;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.AdapterView;
import android.widget.LinearLayout;
import android.widget.Spinner;

import com.elementary.tasks.R;
import com.elementary.tasks.core.utils.LogUtil;
import com.elementary.tasks.core.utils.TimeCount;
import com.elementary.tasks.core.views.roboto.RoboEditText;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class BeforePickerView extends LinearLayout implements TextWatcher {

    private static final String TAG = "BeforePickerView";

    private int seconds = 0;
    private int minutes = 1;
    private int hours = 2;
    private int days = 3;
    private int weeks = 4;

    private RoboEditText mBeforeInput;
    @Nullable
    private InputMethodManager mImm;

    private int mState = minutes;
    private int mRepeatValue;

    public BeforePickerView(Context context) {
        super(context);
        init(context, null);
    }

    public BeforePickerView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public BeforePickerView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        View.inflate(context, R.layout.view_remind_before, this);
        setOrientation(VERTICAL);
        mImm = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        mBeforeInput = findViewById(R.id.before_value_view);
        Spinner beforeType = findViewById(R.id.before_type_view);
        beforeType.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                setState(i);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
        mBeforeInput.addTextChangedListener(this);
        mBeforeInput.setOnFocusChangeListener((v, hasFocus) -> {
            if (mImm == null) return;
            if (!hasFocus) {
                mImm.hideSoftInputFromWindow(mBeforeInput.getWindowToken(), 0);
            } else {
                mImm.showSoftInput(mBeforeInput, 0);
            }
        });
        mBeforeInput.setOnClickListener(v -> {
            if (mImm == null) return;
            if (!mImm.isActive(mBeforeInput)) {
                mImm.showSoftInput(mBeforeInput, 0);
            }
        });
        mRepeatValue = 0;
        mBeforeInput.setText(String.valueOf(mRepeatValue));
        if (attrs != null) {
            TypedArray a = context.getTheme().obtainStyledAttributes(attrs, R.styleable.BeforePickerView, 0, 0);
            try {
                mState = a.getInt(R.styleable.BeforePickerView_before_type, minutes);
            } catch (Exception e) {
                LogUtil.e(TAG, "There was an error loading attributes.", e);
            } finally {
                a.recycle();
            }
        }
        beforeType.setSelection(mState);
    }

    private void setState(int state) {
        this.mState = state;
    }

    private void updateEditField() {
        mBeforeInput.setSelection(mBeforeInput.getText().length());
    }

    public void setBefore(long mills) {
        if (mills == 0) {
            setProgress(0);
            return;
        }
        if (mills % (TimeCount.DAY * 7) == 0) {
            long progress = mills / (TimeCount.DAY * 7);
            setProgress((int) progress);
            setState(weeks);
        } else if (mills % TimeCount.DAY == 0) {
            long progress = mills / TimeCount.DAY;
            setProgress((int) progress);
            setState(days);
        } else if (mills % TimeCount.HOUR == 0) {
            long progress = mills / TimeCount.HOUR;
            setProgress((int) progress);
            setState(hours);
        } else if (mills % TimeCount.MINUTE == 0) {
            long progress = mills / TimeCount.MINUTE;
            setProgress((int) progress);
            setState(minutes);
        } else if (mills % TimeCount.SECOND == 0) {
            long progress = mills / TimeCount.SECOND;
            setProgress((int) progress);
            setState(seconds);
        }
    }

    private void setProgress(int i) {
        mRepeatValue = i;
        mBeforeInput.setText(String.valueOf(i));
        updateEditField();
    }

    private long getMultiplier() {
        if (mState == seconds) return TimeCount.SECOND;
        else if (mState == minutes) return TimeCount.MINUTE;
        else if (mState == hours) return TimeCount.HOUR;
        else if (mState == days) return TimeCount.DAY;
        else if (mState == weeks) return TimeCount.DAY * 7;
        return TimeCount.DAY;
    }

    public long getBeforeValue() {
        long rep = mRepeatValue * getMultiplier();
        LogUtil.d(TAG, "getBeforeValue: " + rep);
        return rep;
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        try {
            mRepeatValue = Integer.parseInt(s.toString());
        } catch (NumberFormatException e) {
            mBeforeInput.setText("0");
        }
    }

    @Override
    public void afterTextChanged(Editable s) {

    }
}

File: app/src/main/java/com/elementary/tasks/core/views/BeforePickerView.kt
package com.elementary.tasks.core.views

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.AdapterView
import android.widget.LinearLayout
import com.elementary.tasks.R
import com.elementary.tasks.core.utils.datetime.DateTimeManager
import com.elementary.tasks.databinding.ViewRemindBeforeBinding
import com.github.naz013.logging.Logger

class BeforePickerView : LinearLayout, TextWatcher, AdapterView.OnItemSelectedListener {

  private lateinit var binding: ViewRemindBeforeBinding

  private var mImm: InputMethodManager? = null
  var onBeforeChangedListener: OnBeforeChangedListener? = null

  private var mState = DateTimeManager.MultiplierType.MINUTE.index
  private var mRepeatValue: Int = 0

  private val multiplier: Long
    get() {
      return when (mState) {
        DateTimeManager.MultiplierType.SECOND.index -> DateTimeManager.SECOND
        DateTimeManager.MultiplierType.MINUTE.index -> DateTimeManager.MINUTE
        DateTimeManager.MultiplierType.HOUR.index -> DateTimeManager.HOUR
        DateTimeManager.MultiplierType.DAY.index -> DateTimeManager.DAY
        DateTimeManager.MultiplierType.WEEK.index -> DateTimeManager.DAY * 7
        DateTimeManager.MultiplierType.MONTH.index -> DateTimeManager.DAY * 30
        else -> DateTimeManager.DAY
      }
    }

  private val beforeValue: Long
    get() {
      val rep = mRepeatValue * multiplier
      Logger.d("getBeforeValue: $rep")
      return rep
    }

  constructor(context: Context) : super(context) {
    init(context, null)
  }

  constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    init(context, attrs)
  }

  constructor(
    context: Context,
    attrs: AttributeSet,
    defStyle: Int
  ) : super(context, attrs, defStyle) {
    init(context, attrs)
  }

  private fun init(context: Context, attrs: AttributeSet?) {
    View.inflate(context, R.layout.view_remind_before, this)
    orientation = VERTICAL
    binding = ViewRemindBeforeBinding.bind(this)

    mImm = context.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager?

    binding.beforeTypeView.onItemSelectedListener = this
    binding.beforeValueView.addTextChangedListener(this)
    binding.beforeValueView.setOnFocusChangeListener { _, hasFocus ->
      if (mImm == null) return@setOnFocusChangeListener
      if (!hasFocus) {
        mImm?.hideSoftInputFromWindow(binding.beforeValueView.windowToken, 0)
      } else {
        mImm?.showSoftInput(binding.beforeValueView, 0)
      }
    }
    binding.beforeValueView.setOnClickListener {
      if (mImm == null) return@setOnClickListener
      if (!mImm!!.isActive(binding.beforeValueView)) {
        mImm?.showSoftInput(binding.beforeValueView, 0)
      }
    }
    mRepeatValue = 0
    binding.beforeValueView.setText(mRepeatValue.toString())
    if (attrs != null) {
      val a = context.theme.obtainStyledAttributes(attrs, R.styleable.BeforePickerView, 0, 0)
      try {
        mState = a.getInt(
          /* index = */ R.styleable.BeforePickerView_before_type,
          /* defValue = */ DateTimeManager.MultiplierType.MINUTE.index
        )
      } catch (e: Exception) {
        Logger.d("init: ${e.message}")
      } finally {
        a.recycle()
      }
    }
    binding.beforeTypeView.setSelection(mState)
  }

  private fun setState(state: Int) {
    this.mState = state
    onBeforeChangedListener?.onChanged(beforeValue)
  }

  private fun updateEditField() {
    binding.beforeValueView.setSelection(binding.beforeValueView.text.toString().length)
  }

  fun setBefore(mills: Long) {
    if (mills == 0L) {
      setProgress(0)
      return
    }
    val beforeTime = DateTimeManager.parseBeforeTime(mills)
    setProgress(beforeTime.value.toInt())
    binding.beforeTypeView.setSelection(beforeTime.type.index)
  }

  private fun setProgress(i: Int) {
    mRepeatValue = i
    binding.beforeValueView.setText(i.toString())
    updateEditField()
  }

  override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
  }

  override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
    try {
      mRepeatValue = Integer.parseInt(s.toString())
      if (mRepeatValue > 0 && s.toString().startsWith("0")) {
        binding.beforeValueView.setText(mRepeatValue.toString())
        binding.beforeValueView.setSelection(binding.beforeValueView.text.toString().length)
        return
      }
    } catch (e: NumberFormatException) {
      binding.beforeValueView.setText("0")
    }
    onBeforeChangedListener?.onChanged(beforeValue)
  }

  override fun afterTextChanged(s: Editable) {
  }

  override fun onNothingSelected(parent: AdapterView<*>?) {
  }

  override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
    setState(position)
  }

  interface OnBeforeChangedListener {
    fun onChanged(beforeMills: Long)
  }
}
